src/main/java/com/fasterxml/jackson/databind/DefaultTyping.java

```diff
@@ -40,13 +40,19 @@
      * "natural" types (String, Boolean, Integer, Double), which
      * can be correctly inferred from JSON; as well as for
      * all arrays of non-final types.
+     * Even if the type is final, when serialized as a root node (not as a field),
+     * default typing is still used because there is no other way to know the type
+     * of the root node when deserialization. for example,
+     * <pre>{@code
+     * objectMapper.writeObjectAsValue(new MyFinalClass()); // even final, still have type information serialized
+     * }</pre>
      * This does NOT apply to {@link TreeNode} and its subtypes.
      */
     NON_FINAL,
 
     /**
      * Value that means that default typing will be used for
-     * all non-final types, with exception of small number of
```

Comment:
Yes, good catch, it should not mention "non-final" there. I will commit this change separately and we can discuss the other part.